Improve Emulator API

aaronemassey opened the pull request:
This change simplifies and reduces the amount of boilerplate
code needed to get a device emulator up and running, thus reducing excise work
to writing tests.

Due to the change in EMUL_DEFINE and other code. This is a breaking-change for emulators.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/39486


Recreation of https://github.com/zephyrproject-rtos/zephyr/pull/45445 due to recreating my fork (ugh I didn't realize it would make my pull-requests go dead).

aaronemassey committed:
emul: Simplify emulator bus initialization and API

Allow emulator creators to write an init function that can be used
across multiple busses so as to reduce the boilerplate and cognitive
load in creating an emulator.

Part of this change includes allowing access to the emul struct from a
field in a {bus}_struct api (e.g. i2c_struct), which removes the need for
sporadic usages of CONTAINER_OF to access the emul struct.

Overall, this change simplifies and reduces the amount of boilerplate
code to get a device emulator up and running, thus reducing excise work
to writing tests.

TEST=twister on accel,espi, and eeprom drivers tests

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
emul: Run clang-format

Run clang-format on all files touched by improved emulator API pull-request
that allowed access to the target device emulator from its bus api without
CONTAINER_OF usage.

drivers/i2c/i2c_emul.c
drivers/spi/spi_emul.c
include/zephyr/drivers/emul.h
include/zephyr/drivers/espi_emul.h
include/zephyr/drivers/i2c_emul.h
include/zephyr/drivers/spi_emul.h
subsys/emul/emul.c
subsys/emul/emul_bmi160.c
subsys/emul/espi/emul_espi_host.c
subsys/emul/i2c/emul_atmel_at24.c

TEST=twister on accel,espi, and eeprom drivers tests

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey committed:
emul: Remove or declare unused args

In several locations of the emulator code there are unused function
arguments that were never caught.

Declare these as unused or remove the unused function parameters entirely.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Had to create a new pull-request due to https://github.com/zephyrproject-rtos/zephyr/pull/45445#issuecomment-1163779915.

Sorry folks.

aaronemassey commented:
Made changes in response to https://github.com/zephyrproject-rtos/zephyr/pull/45445#discussion_r888376932
Did *not* make changes in response to https://github.com/zephyrproject-rtos/zephyr/pull/45445#discussion_r888375548 as author considered it "Fine" for now. The change requested might belong in its own tiny PR.

aaronemassey commented on include/zephyr/drivers/emul.h:
```diff
@@ -18,9 +18,23 @@
+/**
+ * The types of supported buses.
+ */
+enum emul_bus_type {
+	EMUL_BUS_TYPE_I2C,
+	EMUL_BUS_TYPE_ESPI,
+	EMUL_BUS_TYPE_SPI,
+};
```
https://github.com/zephyrproject-rtos/zephyr/issues/47064

aaronemassey committed:
emul: Make emulator parameters consistently named

Some paramaters in prior emulators have historically referred to instances
of struct emul as "emulator" or "emul". The proper parameter name for this
type is "target" as it designates the struct emul instance is the emul
bus (e.g. emul_i2c) controller's target peripheral for dispatch.

Do a small refactor renaming some of these parameters to "target".

TEST=twister on accel, espi, and eeprom drivers tests

Signed-off-by: Aaron Massey <aaronmassey@google.com>

yperess reviewed (dismissed):
Might want to request a squash merge on this since you did the rename in a separate commit. (I'm personally a fan of PRs being squash merged)

galak reviewed (changes requested):
Can we remove the use of `label` from things like emul_get_binding(), espi_emul_register(), etc..

yperess commented:
> Can we remove the use of `label` from things like emul_get_binding(), espi_emul_register(), etc..

Would you suggest something closer to how devices work?

```
#define DEVICE_DT_NAME(node_id) \
	DT_PROP_OR(node_id, label, DT_NODE_FULL_NAME(node_id))
```

aaronemassey committed:
emul: Use DEVICE_DT_GET instead of DT_LABEL

DT nodes aren't guaranteed to define a label property. But emulated bus
controllers currently make use of this property to dispatch to the
associated emulator.

Have emulated bus controllers use DEVICE_DT_GET(node_id) to dispatch to
right target peripheral emulator. This also change makes emul_get_binding
and device_get_binding synonymous in behavior with respect to their
parameters.

This also strictly enforces a 1:1 correspondence between invocations of
DEVICE_DT_DEFINE and EMUL_DEFINE.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
Just added https://github.com/zephyrproject-rtos/zephyr/pull/46818/commits/86603b60c23510173ecc55f6e86637136714a8b5 except for one tiny clang-format that was missed in a prior commit.

aaronemassey commented:
> Can we remove the use of `label` from things like emul_get_binding(), espi_emul_register(), etc..

Now using  DEVICE_DT_NAME(node_id)

aaronemassey commented:
@galak

So using just the `dev` pointer from `DEVICE_DT_GET()` makes a lot of sense in
the context of just developing within the Zephyr project itself.

However, downstream Zephyr developers still converting their legacy drivers to
the Zephyr device driver model that would like to test their drivers using
emulators in the meantime would have difficulty; **this is the case for
ChromiumOS** where we're still porting our drivers to the Zephyr model but we
extensively test them using Zephyr emulators.

In this case, downstream devs have a device node but may only have an
`EMUL_DEFINE()` for that device node and not an associated
`DT_DEVICE_DEFINE()`. Meaning `DEVICE_DT_GET()` would **not** resolve to a
defined address.

Using `DEVICE_DT_NAME()` lets us directly create a 1:1 correspondence between DT
node and emulator, but gives downstream developers the flexibility to test their
legacy drivers with their emulators as they migrate to the Zephyr device model
and upstream their drivers/emulators. Whereas using `DEVICE_DT_GET()` does not
allow for such flexibility as it requires `DT_DEVICE_DEFINE()` to use an
associated emulator.

galak commented:
> @galak
> 
> So using just the `dev` pointer from `DEVICE_DT_GET()` makes a lot of sense in the context of just developing within the Zephyr project itself.
> 
> However, downstream Zephyr developers still converting their legacy drivers to the Zephyr device driver model that would like to test their drivers using emulators in the meantime would have difficulty; **this is the case for ChromiumOS** where we're still porting our drivers to the Zephyr model but we extensively test them using Zephyr emulators.
> 
> In this case, downstream devs have a device node but may only have an `EMUL_DEFINE()` for that device node and not an associated `DT_DEVICE_DEFINE()`. Meaning `DEVICE_DT_GET()` would **not** resolve to a defined address.
> 
> Using `DEVICE_DT_NAME()` lets us directly create a 1:1 correspondence between DT node and emulator, but gives downstream developers the flexibility to test their legacy drivers with their emulators as they migrate to the Zephyr device model and upstream their drivers/emulators. Whereas using `DEVICE_DT_GET()` does not allow for such flexibility as it requires `DT_DEVICE_DEFINE()` to use an associated emulator.

I'm not sure I get this.  `EMUL_DEFINE` already utilizes devicetree as it takes a `node_id` param.  So what are downstream users passing for that param?

aaronemassey commented:
> I'm not sure I get this. `EMUL_DEFINE` already utilizes devicetree as it takes
> a `node_id` param. So what are downstream users passing for that param?

So using `node_id` is completely fine. But the `const device *` pointer from
`DEVICE_DT_GET()` will not resolve to anything if there isn't also a
`DT_DEVICE_DEFINE()` call for the same `node_id`. However, `DEVICE_DT_NAME()`
will resolve as long as the the `node_id` exists.

The specific case I'm discussing is:

There's a `node_id` in a codebase for a peripheral that has a legacy non-Zephyr
driver (this happens in ChromiumOS). Since the legacy driver isn't ported to
Zephyr device driver model yet, it has no `DT_DEVICE_DEFINE()`. **But** we still
want to test the legacy driver using Zephyr emulators. In this case, using
`DEVICE_DT_GET()` in the `EMUL_DEFINE()` macro will break.

Let me know if you'd like to see a ChromiumOS example. We can also do a short
call via Discord or otherwise if you'd like.

galak commented:
> So using `node_id` is completely fine. But the `const device *` pointer from `DEVICE_DT_GET()` will not resolve to anything if there isn't also a `DT_DEVICE_DEFINE()` call for the same `node_id`. However, `DEVICE_DT_NAME()` will resolve as long as the `node_id` exists.
> 
> The specific case I'm discussing is:
> 
> There's a `node_id` in a codebase for a peripheral that has a legacy non-Zephyr driver (this happens in ChromiumOS). Since the legacy driver isn't ported to Zephyr device driver model yet, it has no `DT_DEVICE_DEFINE()`. **But** we still want to test the legacy driver using Zephyr emulators. In this case, using `DEVICE_DT_GET()` in the `EMUL_DEFINE()` macro will break.
> 
> Let me know if you'd like to see a ChromiumOS example. We can also do a short call via Discord or otherwise if you'd like.

But you are already breaking downstream drivers by changing `EMUL_DEFINE()`.  So not sure I see how its that difficult to change those drivers to use `DT_DEVICE_DEFINE` vs changing them to use this new version of `EMUL_DEFINE()`.  And if that is a burden, than change `EMUL_DEFINE` to take a dev pointer and we can put the burden on the caller.

jfischer-no commented on drivers/i2c/i2c_emul.c:
```diff
@@ -74,8 +74,8 @@ static int i2c_emul_get_config(const struct device *dev, uint32_t *dev_config)
 
-static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs,
-			     uint8_t num_msgs, uint16_t addr)
+static int i2c_emul_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
```
Why are you changing this perfectly formatted piece of code here (and on other places)? These changes are not necessary at all.

aaronemassey replied:
[Zephyr uses clang-format](https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format) to properly format code. My understanding is the project is working on making this more consistent and enforced on future changes.

Making code consistently formatted is necessary to improve readability and future diffs easier to parse.

So this change is certainly necessary. Are you suggesting it belongs in its own PR?

jfischer-no replied:
No. From [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style), "The [clang-format tool](https://clang.llvm.org/docs/ClangFormat.html) can be helpful to quickly reformat large amounts of new source code". Is is not new code., most of the code is perfectly formatted. It conform to our coding style, which ->` In general, follow the [Linux kernel coding style](https://kernel.org/doc/html/latest/process/coding-style.html),` -> "The preferred limit on the length of a single line is 80 columns."

aaronemassey replied:
Then it looks like the Code Style section of the contribution guidelines doc is out-of-date. The [clang-format configuration change increasing the line length to 100](https://github.com/zephyrproject-rtos/zephyr/pull/46952) strongly suggests this is the case since there's no sensible reason to have two different styles within the same project unless the project is slowly migrating to from one to the other; that is my understanding here.

[This was the last change to the contribution guidelines](https://github.com/zephyrproject-rtos/zephyr/commit/c7e4388b1dc5a42f171b87af70d2c033b4d9b917) which happened before the `clang-format` config change, which probably should've happened in the [clang-format config change PR](https://github.com/zephyrproject-rtos/zephyr/pull/46952).

Between the contribution guidelines or the clang-format configuration, I'd absolutely consider the clang-format configuration as the source of truth since docs grow stale.

As for applying `clang-format` to "new source code": here `clang-format` was only applied to files changed within this PR. It'd be quite tedious to only apply `clang-format` specifically to new code regions and this would make the code within a source file inconsistently formatted.

Requesting a review from @nashif to comment on this here.

jfischer-no replied:
None of this justifies the change to the perfectly formatted code. I do not see different styles, I see imperfection of clang-format. Also, our coding styles are not so complicated that an average contributor could not follow them.

yperess replied:
Marking this as resolved, @jfischer-no I understand your concern, but the issue here is that if files are formatted differently, then when the next developer is going in and making future changes they will first have to learn the style of the current file in order to match it. The alternative is to have sections of code with different formats (which I think all of us will agree is not acceptable). @aaronemassey 's approach here is reasonable, running clang-format in a separate commit first, then applying changes to make review easier.

jfischer-no replied:
No I strongly disagree. We do not refactor the code in the project that complies with coding style. clang-format reformatted code does not match our established style. And here currently you are enforcing different coding style.

nashif replied:
https://github.com/zephyrproject-rtos/zephyr/issues/30426

jfischer-no replied:
> #30426

Well maximum length in checkpatch does not apply here (and coding-style is already quoted above). The issue is also how clang-format formats the parameters and arguments of a function. There will always be unnecessary changes, and so clang-format is not suited to do formatting without manual post-processing.

jfischer-no reviewed (dismissed):
This PR changes parts of perfectly formatted code, that is not necessary and should be dropped from the PR.

aaronemassey commented:
@galak 

So after some further discussion with @yperess we're going to go with what you've suggested and use `DT_DEVICE_GET()` in place of the node name. I was concerned about stubbing out device structs downstream but it seems easy enough to just stub them out for emulators that don't have a corresponding `DT_DEVICE_DEFINE()`. There's actually an example in this commit so we could get the espi_host tests passing.

I'd prefer to not add `node_id` as a parameter in the `EMUL_DEFINE` for now, but I have documented a required corresponding `DT_DEVICE_DEFINE()`.

Maybe having the stubbed out device structs will be a good reminder for us downstream folks to migrate our legacy drivers to the Zephyr model too :)

By the way, thank you for the time you've spent reviewing this PR!

galak reviewed (commented):
Is there any reason to keep `emul_get_binding` vs replacing it with a version that does the matching based on a `struct device`?

aaronemassey commented:
> Is there any reason to keep `emul_get_binding` vs replacing it with a version that does the matching based on a `struct device`?

Just to keep it consistent with the device_get_binding signature. I don't feel particularly strong about this one, but @yperess does I believe.

aaronemassey commented:
@galak

Keeping name as the parameter for `emul_get_binding()` prevents a required
`device_get_binding()` beforehand which iterates over the devices. It isn't a
significant increase in time iterating, but if clients have to call
`device_get_binding()` for every `emul_get_binding()` it may slightly add up on
some future integration tests that make use of many emulators, which we
certainly have downstream. It's really just we'd have to iterate through all
devices each time when we could just be iterating on emulators.

Let me know if you still don't feel this is a good enough reason and I can make
a change, but this is the other significant reasoning for keeping
`emul_get_binding(name)`.

galak commented:
> @galak
> 
> Keeping name as the parameter for `emul_get_binding()` prevents a required `device_get_binding()` beforehand which iterates over the devices. It isn't a significant increase in time iterating, but if clients have to call `device_get_binding()` for every `emul_get_binding()` it may slightly add up on some future integration tests that make use of many emulators, which we certainly have downstream. It's really just we'd have to iterate through all devices each time when we could just be iterating on emulators.
> 
> Let me know if you still don't feel this is a good enough reason and I can make a change, but this is the other significant reasoning for keeping `emul_get_binding(name)`.

I'm not sure I follow the need to call `device_get_binding`.  We can deal with this in a follow up PR.  My main point is that doing name comparisons doesn't make sense to me when we have `struct device` pointers and we can just do pointer comparisons.  Less costly to compare a pointer than doing a string compare, and more robust to not depend on names.

galak commented on drivers/espi/espi_emul.c:
```diff
@@ -212,41 +212,31 @@ int espi_emul_register(const struct device *dev, const char *name,
-		.receive_vwire = espi_emul_receive_vwire,
-		.manage_callback = espi_emul_manage_callback
-	},
+	.espi_api = { .config = espi_emul_config,
```
nit: formatting here looks odd.  the `.config...` seems like it should be on its own line.

aaronemassey replied:
This is from [a clang-format](https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format), I tried putting the `.config...` on its own line and `clang-format` just puts it back where it is now.
